core/domain/interaction_validation_jobs_one_off.py

```diff
@@ -0,0 +1,278 @@
+                        state.interaction.answer_groups):
+                    for rule_index, rule_spec in enumerate(
+                            answer_group.rule_specs):
+                        if rule_spec.inputs['x'] >= choices_length:
+                            yield (
+                                item.id,
+                                'State name: %s, AnswerGroup: %s,' % (
+                                    state_name.encode('utf-8'),
+                                    answer_group_index) +
+                                ' Rule: %s is invalid.' % (rule_index) +
+                                '(Indices here are 0-indexed.)')
+
+    @staticmethod
+    def reduce(key, values):
+        yield (key, values)
+
+
+class ItemSelectionInteractionOneOffJob(jobs.BaseMapReduceOneOffJobManager):
+    """Job that produces a list of (exploration, state) pairs that use the item
+    selection interaction and that have rules that do not match the answer
+    choices. These probably need to be fixed manually.
+    """
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [exp_models.ExplorationModel]
+
+    @staticmethod
+    def map(item):
+        if item.deleted:
+            return
+
+        exploration = exp_fetchers.get_exploration_from_model(item)
+        for state_name, state in exploration.states.items():
+            if state.interaction.id == 'ItemSelectionInput':
+                choices = (
+                    state.interaction.customization_args['choices']['value'])
+                for group in state.interaction.answer_groups:
+                    for rule_spec in group.rule_specs:
+                        for rule_item in rule_spec.inputs['x']:
+                            if rule_item not in choices:
+                                yield (
+                                    item.id,
+                                    '%s: %s' % (
+                                        state_name.encode('utf-8'),
+                                        rule_item.encode('utf-8')))
+
+    @staticmethod
+    def reduce(key, values):
+        yield (key, values)
+
+
+class InteractionRTECustomizationArgsValidationOneOffJob(
+        jobs.BaseMapReduceOneOffJobManager):
+    """One-off job for validating all the customizations arguments of
+    Rich Text Components.
+    """
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [exp_models.ExplorationModel]
+
+    @staticmethod
+    def map(item):
+        if item.deleted:
+            return
+        err_dict = {}
+
+        try:
+            exploration = exp_fetchers.get_exploration_from_model(item)
+        except Exception as e:
+            yield (
+                'Error %s when loading exploration'
+                % python_utils.UNICODE(e), [item.id])
+            return
+
+        html_list = exploration.get_all_html_content_strings()
+        err_dict = html_validation_service.validate_customization_args(
```

Comment:
The name of this job says it's only about the `Interaction` but I see we are doing it for the entire exploration, am I missing something?

Reply:
Hmm yes, nice catch. Updated the name and moved to exp jobs.